File: src/main/java/core/scm/specialforms/Assert.java
package core.scm.specialforms;

import core.environment.Environment;
import core.evaluator.Evaluator;
import core.exceptions.IllegalSyntaxException;
import core.scm.MutableString;
import core.utils.Utils;

import java.util.List;

public enum Assert implements ISpecialForm {
  ASSERT;

  private static final StackTraceElement[] EMPTY = new StackTraceElement[0];

  @Override
  public Object eval(List<Object> expression, Environment env, Evaluator evaluator) {
    if (expression.size() < 2 || expression.size() > 3) {
      throw IllegalSyntaxException.of(toString(), expression);
    }
    Object result = evaluator.eval(expression.get(1), env);
    if (!Utils.INSTANCE.toBoolean(result)) {
      String message = "";
      if (expression.size() == 3) {
        if (!(expression.get(2) instanceof String) && !(expression.get(2) instanceof MutableString)) {
          throw IllegalSyntaxException.of(toString(), expression);
        }
        message = ": " + expression.get(2).toString();
      }
      AssertionError assertionError = new AssertionError("assert failed" + message);
      assertionError.setStackTrace(EMPTY);
      throw assertionError;
    }
    return Boolean.TRUE;
  }

  @Override
  public String toString() {
    return "assert";
  }
}

File: src/main/java/core/scm/specialforms/Assert.kt
package core.scm.specialforms

import core.Evaluator
import core.exceptions.IllegalSyntaxException
import core.utils.Utils
import core.Writer

object Assert : SpecialForm("assert") {

    private val EMPTY = emptyArray<StackTraceElement>()

    override fun eval(form: List<Any?>, evaluator: Evaluator): Any {
        val message = when (form.size) {
            2 -> "assert failed"
            3 -> "assert failed: \"${evaluator.eval(form[2]) as? CharSequence ?:
                 throw IllegalSyntaxException(toString(), Writer.write(form))}\""
            else -> throw IllegalSyntaxException(toString(), Writer.write(form))
        }
        if (!Utils.toBoolean(evaluator.eval(form[1]))) {
            throw AssertionError(message).apply { stackTrace = EMPTY }
        }
        return true
    }
}
